core/src/main/java/org/apache/spark/ExecutorPlugin.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark;
+
+import org.apache.spark.annotation.DeveloperApi;
+
+/**
+ * A plugin which can be automaticaly instantiated within each Spark executor.  Users can specify
+ * plugins which should be created with the "spark.executor.plugins" configuration.  An instance
+ * of each plugin will be created for every executor, including those created by dynamic allocation,
+ * before the executor starts running any tasks.
+ *
+ * The specific api exposed to the end users still considered to be very unstable.  We will
+ * *hopefully* be able to keep compatability by providing default implementations for any methods
+ * added, but make no guarantees this will always be possible across all spark releases.
+ *
+ * Spark does nothing to verify the plugin is doing legitimate things, or to manage the resources
+ * it uses.  A plugin acquires the same privileges as the user running the task.  A bad plugin
+ * could also intefere with task execution and make the executor fail in unexpected ways.
+ */
+@DeveloperApi
+public interface ExecutorPlugin {
```

Comment:
I think it would be better to have explicit `start()` and `stop()` methods in this interface. Otherwise, especially in the latter case, there is no way for the plugin to know when to do cleanup.

Reply:
Pinging @squito for this one, not sure if he specifically wanted an empty API to start (and have plugins run to their own completion or until the JVM cleans them up) or was considering adding methods later

Reply:
I definitely think a stop would be good.  Since we don't have this currently in a separate thread not sure how start is different from init?  I guess if we do it later or something but think that could be added later unless someone has explicitly use case

Reply:
Both your arguments make sense, I can see a clear purpose for having a `stop`, given that we mostly care about it for successful runs without any weird early failures/terminations. The only case I can make for `start` (if we already have `init`) is if Spark itself calls the `start` methods from a separate thread instead of having plugin writers do it, but it might actually give them more flexibility to do it themselves. Having `start`/`stop` might be better than `init`/`stop`, though.